### PR TITLE
[MM-24146] Add unix socket listener for mmctl local mode

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
This PR exposes `apiv4` through a unix socket.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-24146

### CODE REVIEW

**Comment by mgdelacroix** [comment]
We need to manage the lifecycle of the socket file, as trying to start the server with an existing file in the path will cause an error:

```
Error: Error starting server, err:listen unix /var/tmp/mattermost_local.socket: bind: address alread
y in use: listen unix /var/tmp/mattermost_local.socket: bind: address already in use
Usage:                                                                                              
  mattermost [flags]
  mattermost [command]

Available Commands:
  channel     Management of channels
  command     Management of slash commands
  config      Configuration
  export      Export data from Mattermost
  group       Management of groups
  help        Help about any command
  import      Import data.
  integrity   Check database data integrity
  jobserver   Start the Mattermost job server
  ldap        LDAP related utilities
  license     Licensing commands
  logs        Display logs in a human-readable format
  permissions Management of the Permissions system 
  plugin      Management of plugins
  reset       Reset the database to initial state
  roles       Management of user roles
  sampledata  Generate sample data
  server      Run the Mattermost server
  team        Management of teams
  user        Management of users
  version     Display version information
  webhook     Management of webhooks

Flags:
  -c, --config string        Configuration file to use. (default "config.json")
      --disableconfigwatch   When set config.json will not be loaded from disk when the file is chan
ged.
  -h, --help                 help for mattermost

Use "mattermost [command] --help" for more information about a command.

exit status 1
```

We can either always check&delete the socket file if exists when starting, or maybe better just remove it when shutting down the server and just checking, erroring and not starting if the file exists already at start time 0/5

File: model/config.go [hunk 1]
```diff
@@ -707,6 +709,14 @@ func (s *ServiceSettings) SetDefaults(isUpdate bool) {
 			s.EnableLatex = NewBool(false)
 		}
 	}
+
+	if s.EnableLocalMode == nil {
+		s.EnableLocalMode = NewBool(false)
+	}
+
+	if s.LocalModeSocketLocation == nil {
+		s.LocalModeSocketLocation = NewString("/var/tmp/mattermost_local.socket")
```
Comment (mgdelacroix) [inline model/config.go:718]: This can be a `const` at the beginning of the file and we can import it in `mmctl` for the default path

Comment (isacikgoz) [reply model/config.go:718]: Yes, definitely!

File: cmd/mattermost/commands/server.go [hunk 2]
```diff
@@ -82,6 +82,14 @@ func runServer(configStore config.Store, disableConfigWatch bool, usedPlatform b
 		return serverErr
 	}
 
+	if *configStore.Get().ServiceSettings.EnableLocalMode {
```
Comment (jespino) [inline cmd/mattermost/commands/server.go:85]: Shouldn't it be part of the Server initialization and shutdown?

File: model/config.go [hunk 3]
```diff
@@ -212,6 +212,8 @@ const (
 	OFFICE365_SETTINGS_DEFAULT_AUTH_ENDPOINT     = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize"
 	OFFICE365_SETTINGS_DEFAULT_TOKEN_ENDPOINT    = "https://login.microsoftonline.com/common/oauth2/v2.0/token"
 	OFFICE365_SETTINGS_DEFAULT_USER_API_ENDPOINT = "https://graph.microsoft.com/v1.0/me"
+
+	LOCAL_MODE_SOCKET_PATH = "/var/tmp/mattermost_local.socket"
```
Comment (jespino) [inline model/config.go:216]: Yes, same concern here, why this is `/var/tmp` instead of `/var/run`?

Comment (isacikgoz) [reply model/config.go:216]: It requires `sudo` privileges to create/delete files in `/var/run`. At least for macOS.

File: cmd/mattermost/commands/server.go (hunk 2 above)
Comment (isacikgoz) [reply cmd/mattermost/commands/server.go]: Good point, moved there. 👍

**Pull request merged**